src/backend/gl/src/device.rs

```diff
@@ -1096,38 +1121,56 @@ impl d::Device<B> for Device {
 
     unsafe fn unmap_memory(&self, memory: &n::Memory) {
         let gl = &self.share.context;
-        let buffer = match memory.first_bound_buffer.get() {
-            None => panic!("No buffer has been bound yet, can't map memory!"),
-            Some(other) => other,
-        };
-        let target = glow::PIXEL_PACK_BUFFER;
+        let buffer = memory.buffer.expect("cannot unmap DEVICE_LOCAL memory");
 
-        gl.bind_buffer(target, Some(buffer));
+        gl.bind_buffer(memory.target, Some(buffer));
 
         if self.share.private_caps.emulate_map {
-            let raw = memory.emulate_map_allocation.replace(std::ptr::null_mut());
-            let mapped = slice::from_raw_parts_mut(raw, memory.size as usize);
+            let ptr = memory.emulate_map_allocation.replace(None).unwrap();
+            let allocation = slice::from_raw_parts_mut(ptr, memory.size as usize);
             // TODO: Access
-            gl.buffer_data_u8_slice(target, mapped, glow::DYNAMIC_DRAW);
-            let _ = *Box::from_raw(raw);
+            gl.buffer_data_u8_slice(memory.target, &allocation, glow::DYNAMIC_DRAW);
+            let _ = Box::from_raw(allocation);
         } else {
-            gl.unmap_buffer(target);
+            gl.unmap_buffer(memory.target);
         }
 
-        gl.bind_buffer(target, None);
+        gl.bind_buffer(memory.target, None);
 
         if let Err(err) = self.share.check() {
             panic!("Error unmapping memory: {:?} for memory {:?}", err, memory);
         }
     }
 
-    unsafe fn flush_mapped_memory_ranges<'a, I, R>(&self, _: I) -> Result<(), d::OutOfMemory>
+    unsafe fn flush_mapped_memory_ranges<'a, I, R>(&self, ranges: I) -> Result<(), d::OutOfMemory>
     where
         I: IntoIterator,
         I::Item: Borrow<(&'a n::Memory, R)>,
         R: RangeArg<u64>,
     {
-        warn!("memory range invalidation not implemented!");
+        let gl = &self.share.context;
+
+        for i in ranges {
+            let (mem, range) = i.borrow();
+            let buffer = mem.buffer.expect("cannot flush DEVICE_LOCAL memory");
+            gl.bind_buffer(mem.target, Some(buffer));
+
+            let offset = *range.start().unwrap_or(&0);
+            let size = *range.end().unwrap_or(&mem.size) - offset;
+
+            if self.share.private_caps.emulate_map {
+                let ptr = mem.emulate_map_allocation.get().unwrap();
+                let slice = slice::from_raw_parts_mut(ptr.offset(offset as isize), size as usize);;
+                gl.buffer_data_u8_slice(mem.target, slice, glow::DYNAMIC_DRAW);
```

Comment:
should probably use `buffer_sub_data` instead of rewriting the whole buffer

Reply:
Wow, that's awful I'm very sorry!  There *is* no buffer_sub_data in glow, I will add that.

Reply:
done in personal branch, waiting on glow